### PR TITLE
Implement Subscribe API call for BYOC

### PR DESCRIPTION
**Pull request by edwardrf**
- Implement subscribe API call for byoc aws client, using ecsEventsHandler interface to share data from log tailing to the Subscribe api call
- Make collection stream compliant with the server stream interface, keep all channel logic inside collection stream
- Update term package and tail command to output the same log entry in a single log call, so messages will not be interleaved

### CODE REVIEW

File: src/pkg/term/colorizer.go [hunk 1]
```diff
@@ -121,17 +121,45 @@ func output(w *termenv.Output, c Color, msg string) (int, error) {
...
+	}
+	for _, s := range v {
+		n, err := fmt.Fprint(w, s)
+		l += n
+		if err != nil {
+			return l, err
+		}
+	}
+	if canColor {
+		n, err := io.WriteString(w, ResetColorStr)
```
Comment (nullfunc) [inline src/pkg/term/colorizer.go:152]: Do you need to make this block a defer higher up in case the for-loop above exits out of this function early?

File: src/pkg/cli/client/byoc/aws/subscribe.go [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+	if name == "kaniko" {
+		state = kanikoTaskStateMap[evt.LastStatus]
+		if state == defangv1.ServiceState_NOT_SPECIFIED {
+			return // Ignore unknown kaniko task states
+		}
+		if state == defangv1.ServiceState_BUILD_STOPPING {
+			for _, container := range evt.Containers {
+				if container.ExitCode != 0 {
+					state = defangv1.ServiceState_BUILD_FAILED
```
Comment (nullfunc) [inline src/pkg/cli/client/byoc/aws/subscribe.go:67]: nit: can immediately break out of for-loop on detection

File: src/pkg/cli/client/byoc/aws/subscribe.go [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+		}
+	}
+
+	if !slices.Contains(s.services, service) {
+		return
+	}
+
+	if etag != "" && etag != s.etag {
+		return
+	}
```
Comment (nullfunc) [inline src/pkg/cli/client/byoc/aws/subscribe.go:95]: These two checks can be moved to line 58 for quick exit

File: src/pkg/cli/client/byoc/aws/subscribe.go [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+func (s *byocSubscribeServerStream) send(resp *defangv1.SubscribeResponse) {
+	s.closingLock.Lock()
+	defer s.closingLock.Unlock()
+	if s.closed {
+		return
+	}
+	select {
+	case s.ch <- resp:
+	case <-s.ctx.Done():
```
Comment (nullfunc) [inline src/pkg/cli/client/byoc/aws/subscribe.go:189]: is the <-s.ctx.Done() necessary to detect an error occurring on the channel?

File: src/pkg/clouds/aws/ecs/event.go [hunk 5]
```diff
@@ -18,23 +38,297 @@ type ECSDeploymentStateChange struct {
...
+	status := "TASK_" + e.Detail.LastStatus
+	if e.Detail.StoppedReason != "" {
+		status += " " + e.Detail.StoppedReason
+	}
+	return status
+}
+
+func (e *TaskStateChangeEvent) State() defangv1.ServiceState {
+	state := defangv1.ServiceState_NOT_SPECIFIED
+	if e.Detail.LastStatus == "DEACTIVATING" {
```
Comment (lionello) [inline src/pkg/clouds/aws/ecs/event.go:164]: move out to make it more explicit

File: src/pkg/term/colorizer.go (hunk 1 above)
Comment (edwardrf) [reply src/pkg/term/colorizer.go:152]: Good call!

File: src/pkg/cli/client/byoc/aws/subscribe.go (hunk 4 above)
Comment (edwardrf) [reply src/pkg/cli/client/byoc/aws/subscribe.go:83]: This would prevent the goroutine waiting for the "Receive()" call be blocked when the context is cancelled.

File: src/pkg/clouds/aws/ecs/event.go (hunk 5 above)
Comment (edwardrf) [reply src/pkg/clouds/aws/ecs/event.go:168]: Offline discussion:
- ServiceState should not have task oriented states like "TASK_FAIL"
- Which means we need to parse the status string at the cli subscribe level, which is less desirable
As a result: keep the current implementation.

File: src/pkg/cli/client/byoc/aws/byoc.go [hunk 6]
```diff
@@ -199,7 +199,7 @@ func (b *ByocAws) Deploy(ctx context.Context, req *defangv1.DeployRequest) (*def
 	}, nil
 }
 
-func (b *ByocAws) findZone(ctx context.Context, domain, role string) (string, error) {
+func (b *BocAws) findZone(ctx context.Context, domain, role string) (string, error) {
```
Comment (nullfunc) [inline src/pkg/cli/client/byoc/aws/byoc.go:202]: typo: BocAws -> ByocAws

**Review by nullfunc — APPROVED** [review]
LGTM. minor typo fix.

File: src/pkg/term/colorizer.go [hunk 7]
```diff
@@ -158,7 +184,8 @@ func (t *Term) Printlnc(c Color, v ...any) (int, error) {
 }
 
 func (t *Term) Printfc(c Color, format string, v ...any) (int, error) {
-	return outputf(t.stdout, c, format, v...)
+	line := ensureNewline(fmt.Sprintf(format, v...))
```
Comment (lionello) [inline src/pkg/term/colorizer.go:187]: why was this changed?

Comment (edwardrf) [reply src/pkg/term/colorizer.go:187]: Removed the less usefull `outputf` func above

File: src/pkg/term/colorizer.go [hunk 8]
```diff
@@ -121,17 +121,37 @@ func output(w *termenv.Output, c Color, msg string) (int, error) {
...
+			e = err
+		}()
+	}
+
+	n, err := fmt.Fprint(w, v...)
+	l += n
+	if err != nil {
+		return l, err
+	}
+	return l, nil
```
Comment (lionello) [inline src/pkg/term/colorizer.go:154]:
```suggestion
func append(w io.Writer, canColor bool, c Color, v ...any) (l int, e error) {
	if canColor {
		if l, e = io.WriteString(w, termenv.CSI+c.Sequence(false)+"m"); e != nil {
			return
		}
		defer func() {
			n, err := io.WriteString(w, ResetColorStr)
			l += n
			e = err
		}()
	}

	l, e := fmt.Fprint(w, v...)
	return l+n, e
```

**Review by lionello — APPROVED** [review]
just suggestion to simplify append
